src/main/kotlin/org/jqassistant/tooling/intellij/plugin/editor/rules/refactoring/delete/RuleSafeDeleteProcessorDelegate.kt

```diff
@@ -0,0 +1,256 @@
+        askUser: Boolean,
+    ): MutableCollection<PsiElement> {
+        val files = findEligiblePsiFiles(element)
+        val result = mutableSetOf<PsiElement>()
+        for (file in files) {
+            result.addAll(findRefIdUsages(element, file, "providesConcept"))
+            result.addAll(findRefIdUsages(element, file, "requiresConcept"))
+            result.addAll(findRefIdUsages(element, file, "includeConcept"))
+            result.addAll(findRefIdUsages(element, file, "includeConstraint"))
+            result.addAll(findRefIdUsages(element, file, "includeGroup"))
+        }
+        return result
+    }
+
+    /**
+     * Detects usages which are not safe to delete.
+     * TODO when looking at the usages after dialog is shown by IntelliJ, IDE (2023.3) throws an error unrelated to this code
+     *
+     * @param element an element selected for deletion.
+     * @param allElementsToDelete all elements selected for deletion.
+     * @return collection of conflict messages which would be shown to the user before delete can be performed.
+     */
+    override fun findConflicts(
+        element: PsiElement,
+        allElementsToDelete: Array<out PsiElement>,
+    ): MutableCollection<String> {
+        /*
+         * Explicitly check for conflicts in the context of the tag, its parent and the $element to be deleted
+         * */
+        val messages = mutableListOf<String>()
+        val possibleConflicts =
+            allElementsToDelete.filter { (it as? XmlTag?)?.name == "requiresConcept" }.map { it as XmlTag }
+        val noDeletableParents = possibleConflicts.filter { !allElementsToDelete.contains(it.parent) }
+        val relatedConflicts =
+            noDeletableParents.filter { (element as? XmlTag)?.getAttributeValue("id") == it.getAttributeValue("refId") }
+
+        relatedConflicts.forEach {
+            messages.add(
+                "The ${(it.parent as XmlTag).name} <i>${
+                    (it).getAttributeValue(
+                        "refId",
+                    )
+                }</i> <b>is required</b> by concept <i>${
+                    (it.parent as XmlTag).getAttributeValue(
+                        "id",
+                    )
+                }</i> via <i>${it.name}</i> in file <i>${it.containingFile.name}</i>.",
+            )
+        }
+        return messages
+    }
+
+    /**
+     * Called after the user has confirmed the refactoring. Can filter out some of the usages
+     * found by the refactoring. May show UI to ask the user if some of the usages should
+     * be excluded.
+     *
+     * @param project the project where the refactoring happens.
+     * @param usages all usages to be processed by the refactoring.
+     * @return the filtered list of usages, or null if the user has cancelled the refactoring.
+     */
+    override fun preprocessUsages(
+        project: Project,
+        usages: Array<out UsageInfo>,
+    ): Array<UsageInfo> = usages.map2Array { it }
```

Comment:
Nit: would [toArray()](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/-abstract-collection/to-array.html) not be enough here?